Bitwise costing (PLT-8790)

kwxm opened the pull request:
This extends the costing infrastructure to handle the `ByteStringToInteger` and `IntegerToByteString` functions added in #5654.  The CPU usage of these is quadratic at the moment, and that required quite a few additions.  Once we no longer he tavo support GHC 8.10 it's hoped that we can have a linear-time implementation instead:  this won't require any further changes to the costing infrastructure since a linear function can be represented by a quadratic function with a zero coefficient (or more likely a very small coefficient) in degree 2.   Also, `IntegerToByteString` will currently fail if the length of the output bytestring would be greater than 8000 (accurate costing becomes difficult beyond this point).  We may remove that limit when the implementation is updated, and if we do that will require a new semantic variant.  

`IntegerToByteString` takes a size argument that says how long the output should be, and that required another small extension so that the value of that argument is treated as a literal size for memory costing purposes (ie we don't take the memory usage to be the size of that argument, but the argument itself).

I've added costing benchmarks and run them, and added the resulting data to `builtinCostModel.json`.  The benchmarks may still need a little tuning, but that can be done later.

I also fixed a few small things that I ran into related to plugin integration.

kwxm commented on plutus-core/plutus-core/src/PlutusCore/Bitwise/Convert.hs:
I moved this into this directory because that's where similar code for crypto buitlins is.  Maybe stuff like this should have its own directory.

kwxm commented on plutus-core/plutus-core/src/PlutusCore/Builtin/Runtime.hs:
Drive-by typo elimination.

kwxm commented on plutus-core/plutus-core/src/PlutusCore/Builtin/Runtime.hs:
```diff
@@ -21,7 +21,7 @@ import NoThunks.Class
 -- and the result is stored lazily, since it's not supposed to be forced before accounting for the
--- cost of the application. If the cost exceeds the available budget, the evaluator discards the the
+-- cost of the application. If the cost exceeds the available budget, the evaluator discards the
 -- result of the builtin application without ever forcing it and terminates with evaluation
```
Drive-by typo elimination.

kwxm commented on plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs:
```diff
@@ -1801,19 +1803,28 @@ instance uni ~ DefaultUni => ToBuiltinMeaning uni DefaultFun where
+          {- We disallow inputs of more than 8000 bytes because the execution time becomes difficult to predict
+             accurately beyond this point.  This restriction will be removed in a later variant. -}
+          integerToByteStringDenotation b (LiteralByteSize w) n =
+              if w > 8000 || (w == 0 && (sumCostStream . flattenCostRose . memoryUsage $ n) > 1000)
```
This is a bit annoying, but it has to go here because a later implementation may be able to dispense with this check, and then we'll need two variants and two denotations.

kwxm replied:
The `sumCostStream . flattenCostRose . memoryUsage` pattern appears in several places in the codebase ...

kwxm commented on plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemoryUsage.hs:
```diff
@@ -195,6 +207,17 @@ instance ExMemoryUsage BS.ByteString where
+   would mean is that programs which use the empty bytestring as a builtin
+   argument would become a little cheaper.  The difference would be very small
+   though, so it's maybe not worth fixing. -}
+
```
I just noticed this in passing: it's irrelevant to this PR.

effectfully commented on plutus-benchmark/bls12-381-costs/test/9.6/bls12-381-costs.golden:
```diff
@@ -3,143 +3,143 @@ Hash n bytestrings onto G1 and add points
     n     Script size             CPU usage               Memory usage
   ----------------------------------------------------------------------
-    -     562   (3.4%)       606243489   (6.1%)          264656   (1.9%) 
+    -     516   (3.1%)       493258089   (4.9%)           14164   (0.1%) 
```
1.9% to 0.1%? Does something get optimized away? Or does it now fail? Or why such a drastic difference?

kwxm replied:
> 1.9% to 0.1%? Does something get optimized away? Or does it now fail? Or why such a drastic difference?

Good question.  You had me worried, but it's because the computation's chaining together a bunch of bls12-381 builtins and the memory costs of those are pretty small:
```
bls12_381_G1_add                : 18
bls12_381_G1_compress           : 6
bls12_381_G1_equal              : 1
bls12_381_G1_hashToGroup        : 18
bls12_381_G1_neg                : 18
bls12_381_G1_scalarMul          : 18
bls12_381_G1_uncompress         : 18
bls12_381_G2_add                : 36
bls12_381_G2_compress           : 12
bls12_381_G2_equal              : 1
bls12_381_G2_hashToGroup        : 36
bls12_381_G2_neg                : 36
bls12_381_G2_scalarMul          : 36
bls12_381_G2_uncompress         : 36
bls12_381_finalVerify           : 1
bls12_381_millerLoop            : 72
bls12_381_mulMlResult           : 72
```
The CPU costs are a lot bigger, so the proportional change in the total CPU cost is much smaller.

michaelpj replied:
I mean, this is what we'd expect since the DIY version is awful?

effectfully replied:
Great to know, thanks!

effectfully commented on plutus-core/cost-model/data/builtinCostModel.json:
```diff
@@ -204,7 +204,7 @@
     "bls12_381_G2_neg": {
         "cpu": {
-            "arguments": 307813,
+            "arguments": 129165,
```
That's quite a reduction.

kwxm replied:
That's unrelated to this PR.  I updated the benchmarking results for the bls12-381 functions at some point and they've only just made it into `builtinCostModel.json`.  In retrospect I should probably have updated those in a separate PR to avoid confusion.

effectfully commented on plutus-core/cost-model/data/builtinCostModel.json:
```diff
@@ -237,34 +237,51 @@
     },
-    "bls12_381_mulMlResult": {
+    "bls12_381_millerLoop": {
         "cpu": {
-            "arguments": 2544991,
+            "arguments": 401885761,
```
In case somebody is confused like I was, this is `bls12_381_mulMlResult` being replaced with `bls12_381_millerLoop`, not a 160x growth in cost.

effectfully commented on plutus-core/cost-model/data/builtinCostModel.json:
```diff
@@ -237,34 +237,51 @@
+                "c1": 49601,
+                "c2": 237
+            },
+            "type": "quadratic_in_y"
```
Quadratic? Strange. I guess I need to keep reading.

effectfully commented on plutus-core/plutus-core/src/PlutusCore/Builtin/KnownTypeAst.hs:
```diff
@@ -238,6 +238,7 @@ instance KnownTypeAst tyname uni a => KnownTypeAst tyname uni (Emitter a) where
     toTypeAst _ = toTypeAst $ Proxy @a
     {-# INLINE toTypeAst #-}
 
+
```
Please remove the newline.

effectfully commented on plutus-core/plutus-core/src/PlutusCore/Bitwise/Convert.hs:
```diff
@@ -3,7 +3,7 @@
 
 -- | Implementations for conversion primops from 'Integer' to 'ByteString' and back again.
-module PlutusCore.Builtin.Convert (
+module PlutusCore.Bitwise.Convert (
```
Thanks! `Builtin` is for general builtins matter, not for anything Cardano-specific, so this is a helpful change.

effectfully commented on plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs:
```diff
@@ -1800,20 +1812,31 @@ instance uni ~ DefaultUni => ToBuiltinMeaning uni DefaultFun where
+                                  > integerToByteStringMaximumOutputLength)
+              then do
+                emit "byteStringToInteger result too big"
+                evaluationFailure
```
Thank you for adapting to the new way of doing things.

effectfully commented on plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs:
```diff
@@ -1800,20 +1812,31 @@ instance uni ~ DefaultUni => ToBuiltinMeaning uni DefaultFun where
+          integerToByteStringDenotation b (LiteralByteSize w) n =
+              if w > integerToByteStringMaximumOutputLength ||
+                     (w == 0 &&
+                            8 * fromSatInt (sumCostStream . flattenCostRose . memoryUsage $ n)
```
That is a really weird line... The denotation of a builtin depends on costing for one of the arguments? Can we not rely on costing here and instead having some direct byte-handling logic? What if we change the size metric or something.

kwxm replied:
For [other reviewers](https://github.com/IntersectMBO/plutus/pull/5733#issuecomment-1903468249), see this comment later.

effectfully commented on plutus-core/plutus-core/src/PlutusCore/Default/Universe.hs:
```diff
@@ -417,6 +418,21 @@ instance HasConstantIn DefaultUni term => ReadKnownIn DefaultUni term Word8 wher
+    toTypeAst _ = toTypeAst $ Proxy @Integer
+
+-- See Note [Integral types as Integer].
+instance HasConstantIn DefaultUni term => MakeKnownIn DefaultUni term LiteralByteSize where
```
You could probably simply use newtype-deriving for all the three instances?

kwxm replied:
I think I considered that and then decided that it would be easier just to mimic what it does for the other integral types.

effectfully commented on plutus-core/plutus-core/src/PlutusCore/Default/Universe.hs:
```diff
@@ -417,6 +418,21 @@ instance HasConstantIn DefaultUni term => ReadKnownIn DefaultUni term Word8 wher
+    readKnown term = LiteralByteSize <$> readKnown term
+    {-# INLINE readKnown #-}
+
+
```
Nit: two newlines suck.

kwxm replied:
That was probably just accidental, but sometimes I find it useful to have a bit more space to break the code into blocks of related functionality. Maybe in that case I should use a comment rather than blank lines.

effectfully replied:
I usually use a comment indeed. Rarely I use three newline. Two newlines look like accidental noise to me (not enough signal like with three lines), but that's clearly just me being crazy.

effectfully commented on plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/CostingFun/Core.hs:
```diff
@@ -397,17 +437,23 @@ runTwoArgumentModel
+runTwoArgumentModel
+    (ModelTwoArgumentsQuadraticInY f) =
+        lazy $ \_ costs2 ->
+            CostLast $ evaluateOneVariableQuadraticFunction f $ sumCostStream costs2
```
This needs a comment that costing here isn't lazy, so if lazy costing is required for some future quadratic builtin, this piece of code needs to be rewritten.

effectfully commented on plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/CostingFun/Core.hs:
```diff
@@ -437,6 +483,21 @@ runThreeArgumentModel
             scaleLinearly intercept slope costs3
+runThreeArgumentModel
+    (ModelThreeArgumentsQuadraticInZ f) =
+        lazy $ \_ _ costs3 -> CostLast $ evaluateOneVariableQuadraticFunction f $ sumCostStream costs3
```
Same. I guess we need a Note. And we should reference it in old costing functions too. I should've done that myself when I introduced lazy costing. Feel free to do it here or create a ticket for me or yourself.

kwxm replied:
Since we're talking about laziness, I made a ticket for you to do it at PLT-9312.

effectfully replied:
Thanks!

effectfully commented on plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/CostingFun/Core.hs:
```diff
@@ -437,6 +483,21 @@ runThreeArgumentModel
+    (ModelThreeArgumentsLiteralInYOrLinearInZ (OneVariableLinearFunction intercept slope)) =
+        lazy $ \_ costs2 costs3 ->
+            let width = sumCostStream costs2
+            in if width == 0
```
Measuring `width` in costing units is extremely weird in my opinion and further reinforces the fact that our current approach needs revising.

kwxm replied:
> Measuring width in costing units is extremely weird in my opinion

Yes, I hate this.  Sizes in bytes would be much better (but it's tricky to change that at the moment).

effectfully commented on plutus-ledger-api/test/Spec/CostModelParams.hs:
```diff
@@ -32,8 +32,8 @@ tests =
-            223 @=? length v3_ParamNames
-            223 @=? length V3.costModelParamsForTesting
+            233 @=? length v3_ParamNames
+            233 @=? length V3.costModelParamsForTesting
```
Should be golden tests, I think.

kwxm replied:
> Should be golden tests, I think.

Hmmm.  I just extended what was there already.  @bezirg ?

effectfully commented on plutus-tx-plugin/src/PlutusTx/Compiler/Builtins.hs:
```diff
@@ -452,6 +455,9 @@ defineBuiltinTerms = do
 
+    -- Bitwise operations
+    defineBuiltinTerm annMayInline 'Builtins.integerToByteString               $ mkBuiltin PLC.IntegerToByteString
+    defineBuiltinTerm annMayInline 'Builtins.byteStringToInteger               $ mkBuiltin PLC.ByteStringToInteger
```
Should be moved under the pattern matching right above, following the pattern.

kwxm replied:
Thanks.  I'm not sure how that ended up there.

effectfully reviewed (commented):
Haven't reviewed fully yet, will return shortly. 

Looks good to me, but I don't like how the costing function leaks into the denotation and the denotation leaks into the costing function. We should either embrace that the two are intertwined in a way that does not allow us to separate them meaningfully or do something else, the current approach is really weird in my opinion. Though, I don't mind to merge it first before we figure out a better design (if it's possible).

effectfully reviewed (commented):
Finishing the review with no new comments. 

I'm puzzled as to why the new builtins are quadratic, do you happen to have a rendered plot with data points?

kwxm commented:
> I'm puzzled as to why the new builtins are quadratic, do you happen to have a rendered plot with data points?

That was discussed on Slack.  I've sent you some links.

james-iohk commented on plutus-core/plutus-core/src/PlutusCore/Bitwise/Convert.hs:
Line 41 has a `-- TODO: Cost this builtin.` Can this comment be removed now?

kwxm replied:
Yes, thanks.

kwxm commented:
> I don't like how the costing function leaks into the denotation and the denotation leaks into the costing function.

Neither do I.  It occurs to me now that instead of 
```
  (w == 0 &&
        8 * fromSatInt (sumCostStream . flattenCostRose . memoryUsage $ n)
             > integerToByteStringMaximumOutputLength)
```
I could say 
```
 (w == 0 && 
    n >= 173766203193809456599982445949435627061939786100117250547173286503262376022458008465094333630120854338003194362163007597987225472483598640843335685441710193966274131338557192586399006789292714554767500194796127964596906605976605873665859580600161998556511368530960400907199253450604168622770350228527124626728538626805418833470107651091641919900725415994689920112219170907023561354484047025713734651608777544579846111001059482132180956689444108315785401642188044178788629853592228467331730519810763559577944882016286493908631503101121166109571682295769470379514531105239965209245314082665518579335511291525230373316486697786532335206274149240813489201828773854353041855598709390675430960381072270432383913542702130202430186637321862331068861776780211082856984506050024895394320139435868484643843368002496089956046419964019877586845530207748994394501505588146979082629871366088121763790555364513243984244004147636040219136443410377798011608722717131323621700159335786445601947601694025107888293017058178562647175461026384343438874861406516767158373279032321096262126551620255666605185789463207944391905756886829667520553014724372245300878786091700563444079107099009003380230356461989260377273986023281444076082783406824471703499844642915587790146384758051663547775336021829171033411043796977042190519657861762804226147480755555085278062866268677842432851421790544407006581148631979148571299417963950579210719961422405768071335213324842709316205032078384168750091017964584060285240107161561019930505687950233196051962261970932008838279760834318101044311710769457048672103958655016388894770892065267451228938951370237422841366052736174160431593023473217066764172949768821843606479073866252864377064398085101223216558344281956767163876579889759124956035672317578122141070933058555310274598884089982879647974020264495921703064439532898207943134374576254840272047075633856749514044298135927611328433323640657533550512376900773273703275329924651465759145114579174356770593439987135755889403613364529029604049868233807295134382284730745937309910703657676103447124097631074153287120040247837143656624045055614076111832245239612708339272798262887437416818440064925049838443370805645609424314780108030016683461562597569371539974003402697903023830108053034645133078208043917492087248958344081026378788915528519967248989338592027124423914083391771884524464968645052058218151010508471258285907685355807229880747677634789376
```
The number there is 2^8000, so maybe we could just write that instead.  I think I was a bit worried about having such a big number in the source, but maybe it doesn't matter so much.  We could also use `integerLog2 n > 8000`: maybe that'd be better.

michaelpj commented on plutus-core/plutus-core/src/PlutusCore/Default/Universe.hs:
```diff
@@ -417,6 +418,21 @@ instance HasConstantIn DefaultUni term => ReadKnownIn DefaultUni term Word8 wher
+instance KnownTypeAst tyname DefaultUni LiteralByteSize where
+    toTypeAst _ = toTypeAst $ Proxy @Integer
+
+-- See Note [Integral types as Integer].
```
I think this note is wrong. It might need its own note!

kwxm replied:
That's @effectfully 's note, so he should take a look.  Possibly I've added something that means that it no longer makes sense.

effectfully replied:
I think Michael probably means that the Note does not apply here, not that it is wrong per se? The Note talks about how we encode integral types as `Integer` to get automatic bound checking, but here you do something very different: you need the wrapper for costing reasons and all those instances are literally the same as ones for `Integer`. I would therefore `newtype`-derive them and simply avoid referencing any Notes, `newtype`-deriving is enough of a stress that from the builtins point of view `LiteralByteSize` is indistinguishable from `Integer`.

kwxm commented on plutus-core/cost-model/print-cost-model/Main.hs:
```diff
@@ -131,35 +151,44 @@ renderLinearFunction (LinearFunction intercept slope) var =
+     QuadraticInZ          f   -> [ renderQuadraticFunction f "z" ]
+     LiteralInYOrLinearInZ f -> [ "if y==0"
+                                  , printf "then %s" $ renderLinearFunction f "z"
+                                  , printf "else y bytes"
```
Oh, that's true.  In principle you could use it for CPU usage too, but at the moment it's just for dealing with the awkward memory costs for `integerToByteString`.

kwxm commented on plutus-core/cost-model/data/builtinCostModel.json:
```diff
@@ -128,7 +128,7 @@
     "bls12_381_G1_neg": {
         "cpu": {
-            "arguments": 292890,
+            "arguments": 114242,
```
Yes.  I think the earlier figures for these were perhaps generated before we got the new machine, and then at some point I re-generated them but didn't commit the results.  We should re-run the entire thing before the next HF anyway though.  See [this comment](https://github.com/IntersectMBO/plutus/pull/5733#discussion_r1460246687) too.

kwxm commented on plutus-core/cost-model/budgeting-bench/Main.hs:
```diff
@@ -1,8 +1,10 @@
+{-# OPTIONS_GHC -fno-warn-unused-imports #-}
```
Left over from when I was experimenting.

kwxm commented:
I'll move the size checks back into the implementation so that the denotation doesn't have to care.  If we decide to remove the limit at a later date then we'll have to introduce two variants in `Default.Builtin` to preserve the current behaviour for old scripts.

kwxm commented:
@michaelpj @effectfully I've now moved the size checks out of `Default.Builtins` and back into `Bitwise.Conversions" and added some tests, so you might want to take another look.

kwxm commented:
The maximum size for outputs of `integerToByteString` is 8000, but maybe 8192 would be better.

kwxm commented on plutus-core/plutus-core/src/PlutusCore/Bitwise/Convert.hs:
```diff
@@ -25,23 +29,62 @@ import Data.ByteString qualified as BS
+   integerToByteString in Plutus Core so that we can continue to support the
+   current behaviour for old scripts.-}
+integerToByteStringMaximumOutputLength :: Integer
+integerToByteStringMaximumOutputLength = 8000
```
Maybe 8192 would be better.

effectfully replied:
I'd prefer that indeed, because 8000 feels entirely arbitrary and 8K somehow feels less arbitrary, but it's again my craziness is talking. `8192` also gives the reader a hint of what this is about.

kwxm replied:
> I'd prefer that indeed, because 8000 feels entirely arbitrary

Oh well, I suppose so.  I'll have to change a lot of  the conformance tests now too [goes off grumbling].

effectfully commented on plutus-conformance/agda/Spec.hs:
```diff
@@ -37,8 +38,57 @@ agdaEvalUplcProg (UPLC.Program () version tmU) =
+
 failingTests :: [FilePath]
-failingTests = [
+failingTests =
+    [
+     --- byteStringToInteger
+      "test-cases/uplc/evaluation/builtin/semantics/byteStringToInteger/big-endian/all-zeros"
```
It says `failingTests`, but `all-zeros` isn't a failing test? I'm confused.

kwxm replied:
> It says failingTests, but all-zeros isn't a failing test? I'm confused.

These do fail in the metatheory because in this branch the Agda code doesn't know about the new builtins yet.  That'll be fixed in a PR that'll appear shortly (and which I should have done before this one).  In that branch `failingTests` is empty.

kwxm replied:
If you run the `agda-conformance` tests you get errors like
```
          big-endian
            all-zeros: FAIL (expected)
              Agda: unreachable code reached.
              CallStack (from HasCallStack):
                error, called at /blah/blah/blah/plutus-metatheory-0.1.0.0/build/MAlonzo/RTE.hs:44:23 in plutus-metatheory-0.1.0.0-inplace:MAlonzo.RTE (expected failure)

effectfully replied:
I see, thank you for the explanation.

effectfully commented on plutus-conformance/test-cases/uplc/evaluation/builtin/semantics/integerToByteString/little-endian/bounded/maximum-width-zero/maximum-width-zero.uplc:
```diff
@@ -0,0 +1,4 @@
+-- Check that we can encode zero using the maximum width (8000).
+(program 1.0.1
+ [(builtin integerToByteString) (con bool False) (con integer 8000) (con integer 0)]
```
If you're going to make it 8192, don't forget to update this test. And I guess plenty of other ones. So maybe it's easier to just keep 8000.

effectfully reviewed (approved):
Great tests.

zliu41 commented on plutus-core/plutus-core/src/PlutusCore/Bitwise/Convert.hs:
```diff
@@ -25,23 +28,62 @@ import Data.ByteString qualified as BS
-      emit "If you are seeing this, it is a bug: please report this!"
+  -- Check that the requested length does not exceed the limit.  *NB*: if we
+  -- remove the limit we'll still have to make sure that the length fits into an
+  -- Int (so it should be at most 2^29, which is the largest bound that GHC
+  -- guarantees).
```
Yeah, I don't think we need to worry about 32bit platforms.

michaelpj replied:
Yep, this is right.

kwxm replied:
OK, I've removed that and just said it should fit into an `Int`.

zliu41 reviewed (approved):
Great work.

michaelpj commented on plutus-tx-plugin/src/PlutusTx/Compiler/Builtins.hs:
```diff
@@ -331,6 +334,10 @@ defineBuiltinTerms = do
     defineBuiltinTerm annMayInline 'Builtins.bls12_381_G2_compressed_zero $
         PIR.mkConstant annMayInline BLS12_381.G2.compressed_zero
 
+    -- Bitwise operations
+    defineBuiltinTerm annMayInline 'Builtins.integerToByteString $ mkBuiltin PLC.IntegerToByteString
+    defineBuiltinTerm annMayInline 'Builtins.byteStringToInteger $ mkBuiltin PLC.ByteStringToInteger
```
I think we could just make it throw if you redefine the same name. There's really no reason to do that.

kwxm replied:
> I don't think you need these three lines? 

Ooh, well-spotted.  Maybe we should have some tests that make sure that you can actually use all of the builtins in Haskell code: it's quite easy to forget to add something here or in plutus-tx (although here we've added something too often).  

What happens if you define something twice with different semantics here?  Does the later definition override the earlier one?

michaelpj commented on plutus-tx/src/PlutusTx/Builtins.hs:
```diff
@@ -603,13 +603,13 @@ bls12_381_finalVerify a b = fromBuiltin (BI.bls12_381_finalVerify a b)
 -- [CIP-0087](https://github.com/mlabs-haskell/CIPs/tree/koz/to-from-bytestring/CIP-XXXX).
-{-# INLINEABLE integerToByteString #-}
+{-# INLINABLE integerToByteString #-}
 integerToByteString :: Bool -> Integer -> Integer -> BuiltinByteString
```
This is a good point, for the "exposed" version of the builtins we can indeed give a nicer interface again.

kwxm replied:
Yes, that'd be quite helpful.

kwxm replied:
OK, I've made the plutus-tx versions take a `ByteOrder` argument.

kwxm commented:
Oh, I seem to have accidentally merged the conformance tests with this PR when I was trying to do the opposite. That's a bit confusing.

kwxm commented on plutus-core/untyped-plutus-core/test/Evaluation/Builtins/Conversion.hs:
```diff
@@ -403,6 +403,71 @@ i2bCipExamples = [
+            in [
+             -- integerToByteString 0 maxInput = 0xFF...FF
+             testCase (prefix ++ "maximum acceptable input, no length specified") $
+                      let actualExp = mkIterAppNoAnn (builtin () PLC.IntegerToByteString) [
+                                       mkConstant @Bool () endianness,
+                                       mkConstant @Integer () 0,
+                                       mkConstant @Integer () maxAcceptableInput
+                                      ]
```
:+1: Good idea.

kwxm commented on plutus-conformance/test-cases/uplc/evaluation/builtin/semantics/byteStringToInteger/big-endian/all-zeros/all-zeros.uplc:
```diff
@@ -0,0 +1,4 @@
+-- A bytestring consisting entirely of zeros decodes to 0.
+(program 1.0.1
```
Oops.  I'll just make it 1.0.0 since that's what all the others have and this doesn't depend on 1.1.0 behaviour.

kwxm commented on plutus-core/cost-model/budgeting-bench/Benchmarks/Bitwise.hs:
```diff
@@ -0,0 +1,83 @@
+
+-- Make an integer of size n which encodes to 0xFF...FF
+allFF :: Int -> Integer
+allFF n = 256^(8*n) - 1
```
> What is 256? Is it not `2^(8*n) - 1`?

No, that's right.  We're looking for _size_ n here, which means 8*n bytes (thanks again to the infuriating fact that we measure sizes in 8-byte words).  Since that's a number of bytes, the total number of _bits_ is 8*8*n, and 2^(8*8*n) = 256^(8*n) is 1000...000 with 64*n zero bits.  Subtracting 1 from that removes the 1 at the front and changes all the remaining bits to 1, so you've got 64*n 1 bits, or 8*n 0xFF bytes, which is what you want.  Thanks for making me check though: it's very easy to get this kind of thing wrong.

kwxm commented on plutus-tx-plugin/test/AsData/Budget/9.6/patternMatching.uplc.golden:
Oh!  Now it's gone back to what it was before.

kwxm commented:
Right, I'm going to merge this while the going's good.
